components/vision.py

```diff
@@ -1,169 +1,105 @@
-        self.ping()
-        self.nt.flush()
-        data = self.vision_data_entry.getDoubleArray(None)
-        if data is None:
-            return
-
-        if data[3] <= self.last_data_timestamp:
+    def execute(self):
+        # gets vision data from camera
+        results = self.camera.getLatestResult()
+        self.has_target = results.hasTargets()
+        if not self.has_target:
             return
-        self.last_data_timestamp = data[3]
-
-        # add clock offset to vision timestamp
-        self.vision_data = VisionData(
-            data[0], data[1], data[2], data[3] + self.get_clocks_offset()
+        timestamp = wpilib.Timer.getFPGATimestamp() - results.getLatency()
+        target_pitch = math.radians(results.getBestTarget().getPitch())
+        target_yaw = math.radians(results.getBestTarget().getYaw())
+
+        # work out our field position when photo was taken
+        turret_rotation = self.turret.get_angle_at(timestamp)
+        robot_rotation = self.chassis.get_pose_at(timestamp).rotation()
+
+        # angle from the robot to target
+        target_angle = turret_rotation + target_yaw
+        # distnace from camera to middle of goal
+        self.distance = (
+            PhotonUtils.calculateDistanceToTarget(
+                self.CAMERA_HEIGHT, self.TARGET_HEIGHT, self.CAMERA_PITCH, target_pitch
+            )
+            + self.GOAL_RAD
         )
-
-        # Get vision pose estimate
-        # work out where the vision data was taken from based on histories
-        t = self.vision_data.timestamp
-        turret_angle = self.turret.get_angle_at(t)
-        chassis_heading = self.chassis.get_pose_at(t).rotation().radians()
-
-        # Ranges from vision are from centre of goal to the camera
-        # Add the offset to get range to the centre of the turret
-        range = self.vision_data.distance + self.CAMERA_OFFSET
-
         vision_pose = pose_from_vision(
-            range, turret_angle + self.vision_data.angle, chassis_heading
+            self.distance, target_angle, robot_rotation.radians()
         )
-
         self.field_obj.setPose(goal_to_field(vision_pose))
 
         if self.fuse_vision_observations:
             innovation = vision_pose.translation().distance(
                 self.chassis.estimator.getEstimatedPosition().translation()
             )
             # Gate on innovation
-            if self.gate_innovation and innovation > 3.0:
+            if self.gate_innovation and innovation > 5.0:
                 return
-            # Come up with a position std dev from the fitness reported
-            # When the target is near the edge, the estimate of range is worse
-            pos_std_dev = self.max_std_dev * (1.0 - self.vision_data.fitness)
-            # TODO Can we be smarter and find different values for x and y based on robot orientation?
+
+            std_dev = self.max_std_dev * min(
+                1, max(0, scale_value(self.distance, 5, 8, 1, 0.3))
```

Comment:
Are the output standard deviations the wrong way around here? I would think the closer (i.e. more visible) the target is the more confident we'd be in the vision reading.